python/ray/test/test_global_state.py

```diff
@@ -0,0 +1,46 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import time
+
+import ray
+
+
+def setup_module():
+    if not ray.worker.global_worker.connected:
+        ray.init()
+
+        # Finish initializing Ray. Otherwise available_resources() does not
+        # reflect resource use of submitted tasks
+        ray.get(noop.remote())
+
+
+@ray.remote(num_cpus=1)
+def cpu_task():
+    time.sleep(0.5)
+
+
+@ray.remote
+def noop():
+    return
+
+
+class TestAvailableResources(object):
+    def test_no_tasks(self):
+        cluster_resources = ray.global_state.cluster_resources()
+        available_resources = ray.global_state.cluster_resources()
+        assert cluster_resources == available_resources
+
+    def test_cpu_task(self):
+        cluster_resources = ray.global_state.cluster_resources()
+
+        # Check the resource use of a launched task
+        task_id = cpu_task.remote()
+        available_resources = ray.global_state.available_resources()
+        assert cluster_resources["CPU"] == available_resources["CPU"] + 1
```

Comment:
This test seems potentially flakey since it relies on timing. Can we make it more reliable?

Reply:
Do you think polling `available_resources` alongside `ray.wait` with a small timeout might be a good solution here?

Reply:
Instead, how about 

1. Check initial resources.
2. Call `ray.get(f.remote())` which finishes right away.
3. Call `ray.global_state.available_resources()` and make sure it matches the initial resources. Run this in a loop until it passes (or times out after e.g., 10 seconds).
4. Cal `f.remote()` which sleeps for 1000 seconds, so that we know it won't return.
5. Check `ray.global_state.available_resources()` and make sure it shows the reduced resources. Run this in a loop until it passes (or times out after e.g., 10 seconds).

Make sure the test doesn't take too long.

Reply:
Will do, thanks!